python/hail/expr/expression.py

```diff
@@ -1688,21 +1688,32 @@ def get(self, item):
             >>> hl.eval_expr(d.get('Anne'))
             None
 
+            >>> hl.eval_expr(d.get('Anne', 0))
+            0
+
         Parameters
         ----------
         item : :class:`.Expression`
             Key.
+        default : :class:`.Expression`
+            Default value, must be same type as dictionary value.
 
         Returns
         -------
         :class:`.Expression`
             The value associated with `item`, or missing.
         """
-        if not item._type == self._type.key_type:
+        if not item.dtype == self.dtype.key_type:
             raise TypeError("'DictExpression.get' encountered an invalid key type\n"
                             "    dict key type:  '{}'\n"
-                            "    type of 'item': '{}'".format(self._type.key_type, item._type))
-        return self._method("get", self._value_typ, item)
+                            "    type of 'item': '{}'".format(self.dtype.key_type, item._type))
+        res = self._method("get", self._value_typ, item)
+        if default is not None:
+            result_type = unify_types_limited(self.dtype.value_type, default.dtype)
+            if not result_type:
+                raise TypeError("'get' expects parameter 'default' to have the same type as the dictionary value type, found '{}' and '{}'".format(self.dtype, default.dtype))
+            res = hl.or_else(res, default)
```

Comment:
Can't you have a present key but missing value? In that case, this will return the default. I don't think that's the behavior you intend.

Reply:
ugh. I probably need to write this in Scala, then.

Reply:
was hoping to avoid that. good catch, though.

Reply:
in the case where the default is supplied, can't you just do:
```
if hl.is_defined(key):
    res
else:
    default
```
And if no default is given, it'll be faster to just return `res`.